src/main/java/us/ihmc/handsros2/abilityHand/AbilityHand.java

```diff
@@ -0,0 +1,718 @@
+    * Should be called periodically.
+    */
+   public void update()
+   {
+      long nowNanos = System.nanoTime();
+
+      if (previousTimeNanos > 0L)
+      {
+         long deltaNanos = nowNanos - previousTimeNanos;
+         float dt = Math.min(0.1f, deltaNanos * 1.0e-9f);
+
+         update(dt);
+      }
+
+      previousTimeNanos = nowNanos;
+   }
+
+   /**
+    * Need to supply virtual dt in tests.
+    *
+    * @param dt time step in seconds
+    */
+   void update(float dt)
+   {
+      this.dt = dt;
+
+      AbilityHandControlMode currentControlMode = controlMode.getValue();
+
+      if (currentControlMode != null)
+      {
+         switch (currentControlMode)
+         {
+            case POSITION -> updatePositionControl();
+            case VELOCITY -> updateVelocityControl();
+            case GRIP -> updateGripControl();
+         }
+      }
+
+      previousControlMode = currentControlMode;
+   }
+
+   /**
+    * Updates hand in POSITION mode using per-finger trapezoidal steps.
+    * goalPositions are the targets; goalVelocities set each finger's max velocity.
+    */
+   private void updatePositionControl()
+   {
+      setCommandType(AbilityHandCommandType.POSITION);
+
+      for (int actuatorIndex = 0; actuatorIndex < ACTUATOR_COUNT; actuatorIndex++)
+      {
+         float targetPosition = goalPositions.get(actuatorIndex);
+         float maxVelocity = Math.abs(goalVelocities.get(actuatorIndex));
+
+         float commandedPosition = TrapezoidalStep.step(actuatorPositions.get(actuatorIndex),
+                                                        getFingerVelocityDegPerSec(actuatorIndex),
+                                                        targetPosition,
+                                                        maxVelocity,
+                                                        DEFAULT_MAXIMUM_ACCELERATION,
+                                                        dt);
+
+         setCommandValue(actuatorIndex, commandedPosition);
+      }
+   }
+
+   /**
+    * Updates hand to direct velocity control using goalVelocities.
+    */
+   private void updateVelocityControl()
+   {
+      setCommandType(AbilityHandCommandType.VELOCITY);
+      setCommandValues(goalVelocities.toFloatArray());
+   }
+
+   /**
+    * Performs multi-stage grip control, moving fingers sequentially through {@link AbilityHandGrip#stages}.
+    */
+   private void updateGripControl()
+   {
+      AbilityHandControlMode currentControlMode = controlMode.getValue();
+      AbilityHandGrip currentGrip = grip.getValue();
+
+      // Handle entering GRIP mode vs switching grips while already in GRIP
+      if (previousControlMode != AbilityHandControlMode.GRIP)
+      {
+         gripStage = 0;
+         previousGrip = currentGrip;
+      }
+      else if (previousGrip != currentGrip)
+      {
+         // New grip while already in GRIP: restart stage sequence
+         gripStage = 0;
+         previousGrip = currentGrip;
+      }
```

Comment:
Seems like this can be simplified
```suggestion
      // Check if we're starting a new grip
      if (previousControlMode != AbilityHandControlMode.GRIP || previousGrip != currentGrip)
      {
         gripStage = 0;
         previousGrip = currentGrip;
      }
```